libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/RouteCompatibilityCache.kt

```diff
@@ -34,6 +34,19 @@ object RouteCompatibilityCache {
      * Use to put all routes tracked by `MapboxDirectionsSession` to cache (and clear everything else).
      */
     fun setDirectionsSessionResult(routes: List<NavigationRoute>) {
+        // A crash was reported due to an LruCache inconsistency.
+        // The caching implementation is using java.concurrent
+        // classes for synchronization. This method is getting called
+        // from a coroutine upstream. Putting this call on it's own thread
+        // should make the synchronization effective.
+        Thread {
+            run {
+                setDirectionsSessionResultInternal(routes)
+            }
+        }.start()
```

Comment:
I really doubt that changes anything

Reply:
My thinking was that since synchronize is designed to work with threads rather than coroutines I could execute this code on a new thread and the synchronization would be enforced.  If I change the class to use suspend methods and a kotlin mutex internally that will cause the upstream usage to change.  The methods in this class are ultimately called in `MapboxNavigation`.  If you have suggestions please share them. Based on the stack trace reported, the synchronization blocks seem not to be working correctly. The only explanation I could think of was that one of the methods in this class is originating from a coroutine upstream.

Reply:
Any coroutine code is still run on some thread and synchronization works there as usual. The only thing you can't do is call `suspend` functions inside synchronized blocks, but that is checked during compilation

Reply:
For me this looks like an issue with `hashCode` or `equals`, which causes the `LruCache` to run into issues

Reply:
> Any coroutine code is still run on some thread and synchronization works there as usual. The only thing you can't do is call suspend functions inside synchronized blocks, but that is checked during compilation

You might be right but several coroutines can be associated with the same thread. So I thought if I put the operation on an actual new thread the synchronization would be more effective.

> For me this looks like an issue with hashCode or equals, which causes the LruCache to run into issues

Here is the stack trace from the reported issue. I don't think this is a result of a haschCode/equals problem. 

IllegalStateException
android.util.LruCache.sizeOf() is reporting inconsistent results!

android.util.LruCache in trimToSize at line 203
android.util.LruCache in evictAll at line 311
com.mapbox.navigation.base.internal.route.RouteCompatibilityCache in setDirectionsSessionResult at line 38
com.mapbox.navigation.core.directions.session.MapboxDirectionsSession in setNavigationRoutesFinished at line 49
com.mapbox.navigation.core.MapboxNavigation$internalSetNavigationRoutes$2 in invokeSuspend at line 1129
kotlin.coroutines.jvm.internal.BaseContinuationImpl in resumeWith at line 33
kotlinx.coroutines.internal.DispatchedContinuation in resumeUndispatchedWith$kotlinx_coroutines_core at line 256
kotlinx.coroutines.DispatchedTaskKt in resume at line 178
kotlinx.coroutines.DispatchedTaskKt in resumeUnconfined at line 191
kotlinx.coroutines.DispatchedTaskKt in dispatch at line 163
kotlinx.coroutines.CancellableContinuationImpl in dispatchResume at line 474
kotlinx.coroutines.CancellableContinuationImpl in resumeImpl at line 508
kotlinx.coroutines.CancellableContinuationImpl in resumeImpl$default at line 497
kotlinx.coroutines.CancellableContinuationImpl in resumeWith at line 368
com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl$refreshRoute$callback$1 in invoke$lambda-2 at line 267
com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl$refreshRoute$callback$1 in $r8$lambda$S3_XwtbgUvIeW_lYTzEbEd0UjcE
… $1$$InternalSyntheticLambda$3$5ce679cfe3c67248c45a37495894744b0b268c2244e2f7d1b83487200930d94f$1 in invoke
com.mapbox.bindgen.Expected in fold at line 81
com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl$refreshRoute$callback$1 in invoke at line 248
com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl$refreshRoute$callback$1 in invoke at line 245
com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl$refreshRoute$2$1 in run at line 282
android.os.MessageQueue in nativePollOnce
android.os.MessageQueue in next at line 342
android.os.Looper in loopOnce at line 182
android.os.Looper in loop at line 357
android.app.ActivityThread in main at line 8194
java.lang.reflect.Method in invoke
com.android.internal.os.RuntimeInit$MethodAndArgsCaller in run at line 548
com.android.internal.os.ZygoteInit in main at line 957


> Another option could be to use single thread for all the operations and remove synchronisation at all.

 I can look at doing something like that but I can't change the methods in the caching class to suspend methods because upstream the caching class is used `MapboxNavigation` and would require changing some of those methods to suspend which would break SEMVER.

Reply:
I replaced the LRUCache with a ConcurrentHashMap. Please take a look.